### ISSUE TITLE
Support minetest:// custom browser protocol

### ISSUE THREAD

**Issue by tacotexmex** [minetest/minetest#7400]
##### Issue type
 - Feature request

##### Summary
For the sake of ease-of-use, speed and interoperatibility between desktop and the web, Minetest can have its own custom browser protocol interpreter.

Protocol URLs will start with `minetest://` followed by a agreed upon scheme. Useful URLs I can think of:

- `minetest://server/123.456.789` immediately connects to the IP.

- `minetest://mod/3d_armor` installs the 3d_armor mod from the Content DB.

- `minetest://game/mineclone2` installs the game MineClone 2 from the Content DB.

- `minetest://texturepack/pixelperfection` installs the Pixel Perfection texturepack from the Content DB.

I’m sure there are more examples to be found.

This can really make Content DB a tightly integrated component of the Minetest ecosystem. Users can install content from the convenience of a regular browser instead of or in addition to the formspec based browser.

**Comment by nerzhul** [comment]
interesting idea, you expect the client to be launched connecting to this server then, but currently we need authentication, an intermediare form is required to set the authentication on the client

**Comment by tacotexmex** [comment]
You're right, most servers will need password. The client saves the previously used user name so that should be covered. I figure that at least the URL can boot the client (if not open) and enter the server details in the "Join game" tab for easy access.

The bigger win imo is the ability to use Content DB for installing mods and other content, not just browse them. (The in-game browser is horrible, sorry. The formspec system is at fault though, no one else.)

**Comment by rubenwardy** [comment]
This is something I'd like to add for contentdb

**Comment by ClobberXD** [comment]
An issue might arise if people have multiple versions of Minetest installed side-by-side, or even if they have one portable version, since the browser might not know what or where Minetest actually is...

@nerzhul Authentication is only required if a minetest:// URL is used to join or register on a server. And even so, could it possibly be handled by a browser popup?

**Comment by rubenwardy** [comment]
It should be implemented using JavaScript to detect whether the protocol is available

**Comment by tacotexmex** [comment]
Not sure that’s possible. Either way detection is not really needed, right?

**Comment by ClobberXD** [comment]
As an aside: If the MT installer for Windows is still a thing, it could also be used to create a protocol handler during install time

**Comment by tacotexmex** [comment]
For macOS one simply needs to edit the app bundle’s `Info.plist` in [this manner](https://stackoverflow.com/a/471642).

**Comment by red-001** [comment]
previously discussed in #3527.

**Comment by tacotexmex** [comment]
#3527 was before Content DB and the in-app mod Content downloader, it’s usefulness just got a lot clearer thereby.

**Comment by kkabrams** [comment]
I think a better way to do minetest handling of games, mods, and texture packs would be to use a mime-type or file extension instead of using URLs in that way. That way any website that wants to offer a minetest mod can just serve it up the usual .zip using Content-Type: application/x-minetest-mod or a certain file extension that can be associated with a mod installing script.

**Comment by Calinou** [comment]
@kkabrams As far as I know, one can't override the Content-Type from GitHub Releases or ZIP repository downloads.

**Comment by kkabrams** [comment]
Then how about a manifest file that can contain whether the zip file is a mod, texture pack, etc?

**Comment by rubenwardy** [comment]
> Then how about a manifest file that can contain whether the zip file is a mod, texture pack, etc? I'm looking into if there's a standard format for manifest files inside zips atm. (There doesn't appear to be so far.)

The current standard is this:

* if the folder contains init.lua, it is a mod
* if the folder contains modpack.conf or modpack.txt, it is a mod pack
* if the folder contains game.conf, it is a game
* if the folder contains texture_pack.conf, it is a texture pack
* if none of the above, it may still be a texture pack

**Comment by mercorii** [comment]
What's the situation with this issue?

It would be nice to be able to pass information to servers via shareable links. 

For example, if you want to gather up with friends in a selected place, on a specific server, you could create a link that would tell which server to join, and the link could contain additional parameters to be passed over to server, telling server to do server implementation specific things. such as spawn you at the coordinates given as parameters.

## Use case

Bob wants to hold a meetup with friends, on a specific server at a specific location. Some of the friends have never really played minetest - they have the game installed, but only couple of minutes logged in. To make the gathering easy. Bob shares a special minetest url with his friends via IRC. Using the link everyone can easily join the same server at the same location (given that the server would support passing spawn coordinates, and then would do something with that information).

**The link would then need to contain following parts:**

* "scheme" - `minetest:`
* there would be no "authority" part `//<authority>`, as the authority would be minetest itself - (authority is the part in url that goes: `[userinfo@]host[:port]`)
* action (join server, or interact with server), given as "path". - e.g. `join`
* parameters - `?`
  * parameters defined by minetest protocol that every server should understand 
    * information about server with which to interact - `server=<ip or domain>`
  * additional parameters specific to the server implementation 
    * spawn location - `location=<x>,<y>,<z>`

**Url in full would then be:**

```
minetest:join?server=<ip or domain>&location=<x>,<y>,<z>
```

This way, the application `minetest` would be given instructions to "join" a particular server.

Additionally given that the idea of this specific link would be to join a specific server, the domain could be given in the authority part, making minetest act similar to a browser:

```
minetest://<ip or domain>?location=<x>,<y>,<z>
```

---

The above example urls are constructed based on apple's [instructions](https://developer.apple.com/documentation/uikit/inter-process_communication/allowing_apps_and_websites_to_link_to_your_content/defining_a_custom_url_scheme_for_your_app) and the wikipedia page on [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Generic_syntax). I just read them now the first time, so there's a possibility that I'm misinterpreting some parts of URL construction.

**Comment by kkabrams** [comment]
I've been using my own handler script for minetest links for a couple years now. I use links like: minetest://user@server:port/. The port is optional if there's a SRV record or the server uses port 30000. The user is optional because it defaults to the username of the user running the script. Maybe I should build a windows version as an exe that only handles minetest links? (Something you can point your webbrowser at when it asks "what should we run minetest links with?")

**Comment by rubenwardy** [comment]
In terms of the user-facing side, I suggest using a form like `https://play.minetest.net/join/123.12.31.212:30000/`. Android has the ability to automatically open certain http URLs in an app when an app is installed. That page would open the `minetest://` schema if it detects that it's available, otherwise show a help page.

The benefit of this is the ability to show an informational page to users that don't have the app installed. I believe this is how smart links tend to work, I'd need to look into them more

**Comment by mercorii** [comment]
I'm wondering if `minetest:` schema could be used for improving login process in the future? Perhaps it could be used for creating something like a [slack-style magic link](https://hackernoon.com/magic-links-d680d410f8f7) that carries an auth token in it.

Such a magic link would look like: `https://www.mysite.com/magic_link_signup?token=abcd1234`. 

If servers can do an external authentication check, `minetest:` scheme would allow a way for passing the token, without the player being required to fill forms inside minetest. benrob0329 had a related proposal in his comment on another issue: https://github.com/minetest/minetest/issues/8138#issuecomment-573215104, but the problem there, if i understood correctly, was that formspec was not available for use before authentication.

**Comment by appgurueu** [comment]
For Unix, it works [like this](https://unix.stackexchange.com/questions/497146/create-a-custom-url-protocol-handler)

**Comment by mk-pmb** [comment]
Below, how I register my URL handlers on Ubuntu (trusty to focal) with Xfce. To become effective, the MIME applications cache has to be rebuilt. I do this by restarting my X session, but there should be less intrusive ways.
This website sounds like my approach shoud work for all FreeDesktop/XDG environments: https://edoceo.com/sys/xfce-custom-uri-handler
Exec= and Icon= may use absolute paths. Icon= supports PNG and SVG, if a suitable SVG library is installed.

```text
$ which example-url-handler
/home/ubuntu/bin/example-url-handler
$ cat ~/.local/share/applications/example_url_handler.desktop
[Desktop Entry]
Version=1.0
Type=Application
Encoding=UTF-8
Name=URL Handler Example
Comment=
GenericName=
Exec=example-url-handler %u
Icon=utilities-terminal
StartupNotify=true
Terminal=false
Categories=GNOME;GTK;Network;WebBrowser;
MimeType=x-scheme-handler/pop3;x-scheme-handler/pop3s;x-scheme-handler/smtp;

```

That %u seems to be quoted already. I just tested with this dummy input that only vaguely remembers a URL:
`http://áccènt@example.test/?'"$PATH=${PATH}/"' > /dev/null&½¥=Ω`
At first I had only one double quote character, and xdg-open complained about unpaired quotes. When I added the 2nd double quote character, my script was invoked with the input verbatim, except all double quote characters were removed.

**Comment by mk-pmb** [comment]
Just tested on Windows 8.1 Pro, 64 bit:

```ini
REGEDIT4

[HKEY_CLASSES_ROOT\minetest]
@="MineTest Protocol"
"FriendlyTypeName"="MineTest URL"
"URL Protocol"=""

[HKEY_CLASSES_ROOT\minetest\DefaultIcon]
@="notify.exe,0"

[HKEY_CLASSES_ROOT\minetest\shell]
@="open"

[HKEY_CLASSES_ROOT\minetest\shell\open]

[HKEY_CLASSES_ROOT\minetest\shell\open\command]
@="c:\\path\\to\\notify.exe MineTest-URL \"%1\""

```

Make sure the `.reg` file is on a drive that the admin user can access, e.g. not not a virtual drive aliased by a user via the "subst" command. The association becomes effective immediately. You can verify this by creating a new shortcut on the desktop, for a URL as the target. Also I have no idea why the icon program is looked up in PATH, but the command has to be an absolute path. Before testing, I was convinced it would resolve the command as well.

**Comment by mk-pmb** [comment]
Somehow the blank line at the end of the reg file was lost. I have a vague memory that at least on WinXP, it was mandatory, else the last line before it would be ignored.

**Comment by mk-pmb** [comment]
What's left to do here? Can I help make progress?

**Comment by sfan5** [comment]
* A proposal with an exact specification (of what the `minetest://` protocol can do) should be agreed upon<b>(!)</b>
* Someone needs to implement protocol handling into the code
* Someone needs to do the platform-specific support stuff (registry changes, desktop file as seen above) and integrate it
* Ideally a PR is opened that includes both of the above points

Regarding the first point, I think supporting only `minetest://connect/$ip_or_domain/$port` for now is sufficient.
Upon opening, the client should show a prompt asking the user if they want to connect to the given server and also an username/password field.

**Comment by mk-pmb** [comment]
I'd like to use this as an easy interface to login managers, so I'd prefer we use `minetest://[user[:pswd]@]host[:port]/connect/` instead.

**Comment by rubenwardy** [comment]
you can use existing shortcuts or popen/etc for that

**Comment by mk-pmb** [comment]
Also I'd like to skip the connect confirmation if the server is saved as a favorite, and that confirmation dialog should have an option like "Add this server to favorites, so in the future you can connect without this question".

**Comment by sfan5** [comment]
Would it just use whatever name was last entered into the mainmenu? That sounds a bit inconvenient at times.

**Comment by mk-pmb** [comment]
@sfan5 good catch. It should probably auto-connect only if the URL contains a username and a password (even if empty). otherwise, URLs to `/connect` should just pre-fill the join game tab.

**Comment by mk-pmb** [comment]
I'll try and make a prototype as a separate project, and if you like it, we can later package it together with MT.

**Comment by Emojigit** [comment]
I suggest:
* `minetest-connect://FooBar@example.com:30000` for servers
* `minetest-cdb://content.minetest.net/kay27/mineclone5` for CDB (MCL5 for example, this can also be used to download texture packs and mods)

**Comment by appgurueu** [comment]
> I suggest:
> 
>     * `minetest-connect://FooBar@example.com:30000` for servers
> 
>     * `minetest-cdb://content.minetest.net/kay27/mineclone5` for CDB (MCL5 for example, this can also be used to download texture packs and mods)

Why use two protocols if only one is needed?

**Comment by kkabrams** [comment]
> I suggest:
> 
>     * `minetest-connect://FooBar@example.com:30000` for servers
> 
>     * `minetest-cdb://content.minetest.net/kay27/mineclone5` for CDB (MCL5 for example, this can also be used to download texture packs and mods)

minetest-cdb: seems like a bad idea. I'd much rather have content-db stuff use a specific file extension or content-type instead of trying to put that metadata into the URI scheme. maybe .mtz for [m]ine[t]est[z]ip and then peoples' OSs can pass all the .mtz files to minetest, no matter if they got it from http, ftp, or off of a CD or USB drive. maybe content-types of something like, application/minetest-mod, application/minetest-texture-pack, application/minetest-game. I've heard the argument before that this isn't a practical solution because most places that people host these zips don't allow custom filename extensions, or content-types. Well... those places suck in their inflexibility. I dunno.

**Comment by rubenwardy** [comment]
I'd prefer `minetest://content.minetest.net/packages/kay27/mineclone5/`. To begin with, it should be based on the hostname - it should match contentdb_url in the config. As Minetest will need to load the package list and other API requests, there isn't much flexibility on changing the ContentDB URL. Perhaps in the future a metadata file could be used

**Comment by PolySaken-I-Am** [comment]
Necroing because I have an important point in case this continues:

I've spotted a potential issue with the idea of having the database address in the URI; namely, it allows people to create links that cause MT to download and install anything from anywhere. This is a problem for obvious reasons, and it's especially an issue for new users not understanding the safety issue and installing malicious code from somewhere.

This could be partially mitigated with a warning when connecting to a new database, however it's still a major problem.

Also, I would suggest using `minetest://content/<data>` and `minetest://connect/<data>`, where data is a comma-separated list of keyword parameters such as `address=0.0.0.0`. This makes it less likely to be mistaken for a normal URL, and means that the parameters and their positions aren't hardcoded.

**Comment by mk-pmb** [comment]
> it allows people to create links that cause MT to download and install anything from anywhere.

I think that's what @rubenwardy meant with

> it should match contentdb_url in the config.

I understood it as that MT will deny, or at least drastically warn about, any content URLs that do not start with the configured `contentdb_url`.

> This makes it less likely to be mistaken for a normal URL,

What is a "normal" URL? The closest thing I can find in [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986) would be the normalizer, but that doesn't seem to fit.

> and means that the parameters and their positions aren't hardcoded.

I seem to have missed something here. Can you give an example of what would be confusing and what to better write instead?

**Comment by PolySaken-I-Am** [comment]
> I understood it as that MT will deny, or at least drastically warn about, any content URLs that do not start with the configured contentdb_url.
Then why have the DB address in the URL at all, if it will be denied?
> What is a "normal" URL? 
Basically a website URL, like `https://github.com` or example. "Normal" doesn't mean anything other than what most people think of when they hear "URL".
> I seem to have missed something here. Can you give an example of what would be confusing and what to better write instead?
For example, the `minetest-connect://FooBar@example.com:30000` suggestion has a fixed format; username, then hostname, then port. This reduces expandability and expressiveness for little to no benefit.

**Comment by mk-pmb** [comment]
> Then why have the DB address in the URL at all, if it will be denied?

So pro users, or future versions of default MT, can arrange for smarter alternatives than just deny.

> has a fixed format; username, then hostname, then port.

Yes. Those are defined by the URL schema. We're trying to make a URL handler. We should really stick to the URL schema as strictly as we can, or we will run into all kinds of compatibility issues because operating systems will have certain expectation about what an URL is. If we fail to match them, the OS won't apply our URL handler, or will try to "fix" the broken URL.

For the kind of arbitrary named pairs that you suggest, URLs have a query string part.

**Comment by PolySaken-I-Am** [comment]
> Yes. Those are defined by the URL schema...

The format of everything coming after the `minetest://` part is completely up to the handler. Most URL schemes don't comply at all with that format (take `data:text/html,`, or the steam URL format, or anything other than the most basic URL schemes for example). The URL scheme used by minetest won't be recognized by almost anything anyway; short of being able to write markdown, bbcode, or HTML, most sites only consider `http://` and `https://` as valid URLs. The worst case scenario is really that the URL looks like plain text.

> So pro users, or future versions of default MT, can arrange for smarter alternatives than just deny.

Is there an actual usecase for this though? Forks are likely to have their own CDB, and I don't see a practical application for allowing links to nonstandard databases that's worth the security risk. At most, a "database ID" should be definable which tells MT that the requested package isn't on the selected `contentdb_url`.

**Comment by PolySaken-I-Am** [comment]
Should probably mention the difference between URL and URI here. We're talking about a URI handler, not a URL handler. A URL is one specific type of URI, which doesn't concern this discussion.

**Comment by mk-pmb** [comment]
Right. However, a lot of messengers and similar apps will only make your URI (or even IRI) clickable if they resemble a URL enough.

**Comment by PolySaken-I-Am** [comment]
We don't need it to be clickable on most things, to be entirely honest. This tool would be much more useful on the forums and similar than anywhere else.

**Comment by rubenwardy** [comment]
The reason I suggested `minetest://content.minetest.net` is because I assumed that the URI needed to have a valid domain in it. If this is not the case, then `minetest://content/` is fine.

As for portability, I'd suggest wrapping the URLs with a HTTP landing page. So instead of linking directly to `minetest://server.com:123`, you'd link to `https://open.minetest.net/join?address=server.com&port=123` or something. That page would then attempt to open the scheme, or show instructions and a link to download Minetest if it fails

One blocker to this issue would be #9166

**Comment by PolySaken-I-Am** [comment]
is that approach for mobile compatibility or something? I'm not sure I understand the need for a HTTP redirect page.

**Comment by appgurueu** [comment]
> > As for portability, I'd suggest wrapping the URLs with a HTTP landing page.
> 
> Is that approach for mobile compatibility or something? I'm not sure I understand the need for a HTTP redirect page.
> 
> If it's just to provide help for people clicking on the link without minetest installed, then that seems a little unnecessary; other launchers that have custom URI schemes don't do the same because it's pretty much a given that anyone attempting to install a package or join a server already has the application installed.
> 
> (Also this just adds yet another web service that fork developers would need to maintain or remove support for, whereas a normal URI handler doesn't need any work to be adapted for a fork)

It's not that much effort (no "webservice" needed); GH pages probably suffices, as opening the URL can be done all clientside. It's just a convenience as HTTP links are usually handled better than custom, unrecognized protocols - and we can include the install instructions if opening the link should fail.

**Comment by rubenwardy** [comment]
> other launchers that have custom URI schemes don't do the same because it's pretty much a given that anyone attempting to install a package or join a server already has the application installed.

Discord does this. And this assumption is incorrect - imagine a friend sharing a link to join the server they're on

> (Also this just adds yet another web service that fork developers would need to maintain or remove support for, whereas a normal URI handler doesn't need any work to be adapted for a fork)

Not our concern really. But as said, it can just be a GH pages website

> It's not that much effort (no "webservice" needed); GH pages probably suffices, as opening the URL can be done all clientside. It's just a convenience as HTTP links are usually handled better than custom, unrecognized protocols - and we can include the install instructions if opening the link should fail.

Exactly this. Having a wrapper like this makes the links very portable, and 

>  There's also an argument that consistency with other URI schemes is a bad thing in part, because we don't want the minetest URI format to be confused with a different scheme or a normal URL

**Comment by ProgrammerWhoPrograms** [comment]
`https://open.minetest.net/join/server.com:123/` would be a single point of failure AFAICS.
I concur with @mercorii that the URI shouldn't use a '//' if there is no authority component afterwards. I.e. it should use 'minetest:/server/minetest.example' instead of 'minetest://server/minetest.example'.

**Comment by rubenwardy** [comment]
I think that the added ease-of-use makes up for it being a single source of failure. Server websites could do a similar thing themselves instead, open.minetest.net wouldn't have a special status

**Comment by kkabrams** [comment]
How about, both? minetest://user:pass@host:port/ can be used to directly connect, but a "share this server" thing will default to opening a page to some https link that gives information for people that don't have minetest installed yet, and /that/ page can then, have a "open server" button that uses a minetest://user:pass@host:port/ URI? If you want to be able to confirm that a user was able to open the link when they click on it, then have it so the minetest config has a setting to pick which server to send a "the user connected" message to.

**Comment by kkabrams** [comment]
in the case of portable installs, you can make minetest do a check to see if something already has a minetest URI handler set, if not, put a button or popup (yeah, I know, annoying) asking "would you like to set this version of minetest as the default handler for minetest links?"

**Comment by rubenwardy** [comment]
> How about, both? minetest://user:pass@host:port/ can be used to directly connect, but a "share this server" thing will default to opening a page to some https link that gives information for people that don't have minetest installed yet, and /that/ page can then, have a "open server" button that uses a minetest://user:pass@host:port/ URI?

This is exactly what open.minetest.net would need to do. It can't just magically open Minetest, there would still need to be a custom scheme

**Comment by Emojigit** [comment]
I made a prototype of open.minetest.net, now hosted on https://open-minetest.thost3.de.

Try the following:
* Hostname and port only: https://open-minetest.thost3.de/toyshost2.ddns.net:30010
* With username: https://open-minetest.thost3.de/Cato@toyshost2.ddns.net:30010
* With password: https://open-minetest.thost3.de/Cato:A_OF_COURSE_FAKE_PASSWORD@toyshost2.ddns.net:30010

Source code: https://gist.github.com/Emojigit/fcba17fda53cf73967c3fc869aedf088

**Comment by rubenwardy** [comment]
Including a password only makes sense if it's a server password, not your password. Minetest doesn't have such a concept yet

**Comment by GoodClover** [comment]
Gonna bring out the URL pedantic out in me, forever annoys me when I see yet another messed up scheme. Just going to chuck my ideas onto the pile.

The first part of a `/path` is what distinguishes the different actions.

## Join server
`minetest://username:password@host:port/join`
If omitted, prompt the user for a name & password.
*Never* connect automatically, always have at least one button between clicking the URL and joining.

`?query` could be passed to `on_joinplayer`, so a mod can do whatever it likes with it, e.g. TP to a position, take a login token etc.
What goes inside `?query`s are *not* a part of individual URL scheme specs, they are free-form to be sent to the server. Much like how in MT the `param` to a chatcommand is just a string, it's the same here.
The only decision to make is enforcing `k=v&flag&k2=v2` vs just passing a string and providing a util to parse the former.

## View CDB package
`minetest://instance:port/content/username/package`
`username/package` are as in CDB's API.
If instance is omitted, default to the one in `contentdb_url`. Make it *very* clear that you're viewing from a different instance to normal if it's not `contentdb_url`.
Maybe even show a list of all packages on the instance/by a specific user if the username/package is omitted. (separate feature request)

## Send data whilst in-game
`minetest:/intent/name?query`
Anther idea is for links to be able to send data to an already connected server.
This could be used e.g. to link to an in-game mail system, wiki, or to a teleporter pad.
```lua
minetest.register_intent(name, function(query))
```

It's important this one works from the in-game chat as well as from external programs.

**Comment by fgaz** [comment]
My proposal:

## Join server

As per https://github.com/minetest/minetest/issues/7400#issuecomment-1242991699 .

The URL authority is important because it enables stuff like #12943

## CDB packages

Adopt the [purl spec](https://github.com/package-url/purl-spec), where

* **type** is **`minetest`** (or `minetest-cdb` or `minetest-contentdb`. it doesn't matter as long as we choose one)
* **namespace** is the **package author**.
* **name** is the **technical name** of the package.
* **version** is the contentdb release number. The human-readable version is mutable and uniqueness isn't enforced.
* **qualifiers** may contain `repository_url` set to the contentdb domain, defaulting to content.minetest.net

Examples:

* `pkg:minetest/Minetest/minetest_game`
* `pkg:minetest/Minetest/minetest_game@13955`
* `pkg:minetest/example_author/example_mod?repository_url=example.org`

**Comment by cweiske** [comment]
From what I read in the comments, this issue here seeks to solve 4 tasks:
1. Join a server in minetest by clicking on a link in a browser or messaging application
2. Install a minetest mod by clicking on a link in a browser
3. Install a minetest game by clicking on a link in a browser
4. Install a minetest texture pack by clicking on a link in a browser

Since downloading+installing files and joining a server are very separate things, I would handle them very differently.

## About installing packages
When only the official contentdb shall be supported, adding a special command to the "minetest:" protocol would be ok.

This would limit the scope, and I would rather like mods and texture packs to be installable from anywhere - contentdb, github, self-hosted website. For this to work the actual download protocol (ftp, http, https or whatever comes next) needs to be in the URL, and thus "minetest:..." is not sufficient, because the protocol is replaced there.

So for installation URLs, a custom MIME type should be used, and minetest should register itself as handler for that MIME type.
Process:
1. User clicks on file link in the browser
3. Browser downloads file
4. Browser sees MIME type sent by the server and asks user if minetest shall be opened
5. Minetest gets started with local file passed as parameter and installs mod/game/texturepack (maybe with confirmation)

Here the file extension could stay `.zip`, but this would require that the server sends the correct custom minetest mimetype as header in the download response. This is possible on a custom server (and contentdb), but not in most public hosters like github.

An alternative would be to give the files an additional file extension, and let minetest also register itself as handler for the file extension. This would make the handler work for downloads from github and make it easier to offer downloads from own servers, because special mime type configuration for `.zip` files would not be needed.

I suggest the following mime types:
- `application/minetest-game`, extension `.mtgame`
- `application/minetest-mod`, extension `.mtmod`
- `application/minetest-texturepack`, extension `.mttexturepack`


## Joining a server
Joining the server is the only task left here, so we only need to take care of that in the `minetest` protocol.

URIs are built by having a protocol (= scheme), a colon, and the rest is specific for the protocol.
Some standards emerged, and one of them is having the server to connect to as first part in the URL (`http://example.org`, `mailto:user@example.org`).

The main case for join links is enabling strangers to join a server, without the website knowing their name.

I suggest using `minetest:hostname.example.org:30000` or `minetest:23.42.17.18:30000` for join links without a user name. The port number should be optional and default to 30000 if not given.

When a website knows the username of the user looking at the page, the name can be given in the URI just like for e-mail links: `minetest:user@hostname.example.org`.
On the UI side, minetest should of course ask the user if he wants to join, instead of joining automatically.

**Comment by beedell-roke** [comment]
> https://github.com/minetest/minetest/issues/7400#issuecomment-633106720

Summarily, I believe that @mercorii's URI syntax is best.

#### **Variably Positionable**

To explain, imagine invoking a binary from `$PATH` in your CLI shell, using the standard POSIX CLI shell command syntax: [^POSIX_command_syntax]

<table><tr>
    <th>Installation</th>
    <td><code>minetest install --package='package_name'</code></td>
</tr><tr>
    <th>Server Connection</th>
    <td><code>minetest join --server='ip=192.168.0.0,port=3000' --location='x=0,y=0,z=0' --username=example_username --password="$password"</code></td>
</tr></table>

Of course, the actions there could be flags too, but it's more readable in this example otherwise (and the first value is in essence a flag if no following values are hierarchical).

In essence, that is the most versatile that a command structure can be. The answer replicates this:

<table><tr>
    <th>Installation</th>
    <td><code>minetest:install?package='contentdb_package_identifier'</code></td>
</tr><tr>
    <th>Server Connection</th>
    <td><code>minetest:join?host='ip=192.168.0.0,port=3000'&username='example_username'&password=''</code></td>
</tr></table>

<details>

1.	I've used CSV when nested arguments are necessary so that the examples remain readable. Otherwise, they should be percent-encoded.[^encode]

1.	Note that I have adhered to the undermentioned section of the answer:
	
	> action (join server, or interact with server), given as "path". - e.g. join
	
	...because I'm not aware of whether URIs support an empty path (even with the authority absent) yet with an argument section present.

</details>

#### **Statically Positionable**

To compare, alternatives utilize a hierarchical structure: [^example]

<table><tr>
    <th>Installation</th>
    <td><code>minetest:install/contentdb_package_identifier</code></td>
</tr><tr>
    <th>Server Connection</th>
    <td><code>minetest:join/hostname.example.org:30000/example_username</code></td>
</tr></table>

Such a structure shall quite quickly cause problems, should additional functionality be added which necessitates a reorganisation.[^advocacy] It is evidently inferior.

Consider the difference between a path and JSON - mere values, versus named, nestable values.

[^advocacy]: https://webmasters.stackexchange.com/questions/139773/how-to-best-create-urls-for-google-and-other-search-engines-using-query-paramet#comment204020_139773
[^POSIX_command_syntax]: https://softwareengineering.stackexchange.com/revisions/70418/4#:~:text=The%20POSIX%20style%20consists%20of%20options%20with%20a%20single%20dash%20followed%20by%20a%20single%20letter%20indicating%20the%20option%2C%20with%20the%20argument%20value%20separated%20from%20the%20option%20by%20a%20space.
[^example]: https://github.com/minetest/minetest/issues/7400#issuecomment-1325003101
[^encode]: https://stackoverflow.com/revisions/1555074/1#:~:text=Yes%2C%20the%20quotation%20mark%20character%20(U+0022)%20is%20not%20allowed%20in%20plain%20and%20must%20be%20encoded%20with%20%2522.

**Comment by ghost** [comment]
i highly support this idea and would love to being attention back to it. now of course it would be luanti:// or something. but i really support the idea of using this for connecting to servers.

what i would hope it could also do is provide the username and maybe even the password in the url. so like `luanti://ip=example.com&user=bob&password=abc123` or something like that. of course, the latter 2 parameters should be _optional_ but i think they should be available. me in particular, i would use the password parameter for dynamically changing one-time login token type passwords using a custom authentication system. thats actually why i want this custom luanti URI to exist. so that i could have a server where people log into through my website. if they are logged into the website, they can click Play and a one time token is generated and used as their password to log into the luanti server and this URI would make that way more convenient for the user as it could be one click instead of asking them to copy and paste the one time token into the password box when they connect.  and to be clear, all that token stuff i mention is stuff i would design myself. thats _not_ part of this feature request. just saying thats what i would use this feature for. to make tokens more convenient.

thx for reading. i really hope this feature is added. ^^

**Comment by beedell-roke** [comment]
@JasonTable, I agree. Even in personal code, if one utilises a scripting language that provides command substitution, so that the user can utilise an external credential store (like [`bw`][1]), there is no security concern. I think not providing credential parameters would be foolish.

[1]: https://github.com/bitwarden/clients/tree/60855c734fe34ed98ae321777266ededdb571e6d/apps/cli#help-command:~:text=global%20%2D%2Dhelp%20option:-,bw,-%2D%2Dhelp

**Comment by sfan5** [comment]
An important thing to mention is that any design that we actually implement should work on Android. And I don't think you can just register a `protocol://` on Android.

**Comment by rubenwardy** [comment]
You can register custom schemes on android

```
  <intent-filter android:label="@string/filter_view_example_gizmos">
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <!-- Accepts URIs that begin with "example://gizmos” -->
        <data android:scheme="example"
              android:host="gizmos" />
    </intent-filter>
```

**Comment by ghost** [comment]
OOh i also just had the thought that this should also be integrated with the url_button formspec somehow. maybe instead of asking to visit the url, it asks if you want to leave this server and join ____ server. and if username or password is not filled out then it also asks for those in that box.

but eh, not as big of a priority for me as i just want the URI feature to exist. but i figured i would share this idea too.
